corehq/apps/domain/models.py

```diff
@@ -622,6 +623,47 @@ def get_or_create_with_name(cls, name, is_active=False,
             new_domain.save(**get_safe_write_kwargs())
             return new_domain
 
+    @classmethod
+    def generate_name(cls, hr_name, max_length=25):
+        '''
+        Generate a URL-friendly name based on a given human-readable name.
+        Normalizes given name, then looks for conflicting domains, addressing
+        conflicts by adding "-1", "-2", etc. May return None if it fails to
+        generate a new, unique name.
+        '''
+
+        original_name = hr_name.strip().lower()
+        original_name = re.sub(r'[^0-9a-z]+', '-', original_name)
+
+        def has_conflict(name):
+            return Domain.get_by_name(name) or Domain.get_by_name(name.replace('-', '.'))
+
+        name = original_name
+        shorten = -1
+        while has_conflict(name) and shorten + 2 < max_length:
+            shorten = shorten + 1
+            name_prefix = original_name
+            if shorten > 0:
+                name_prefix = name_prefix[:-shorten]
+            hits = Domain.get_names_by_prefix(name_prefix + "-")
+            if re.search(r'\.', name_prefix):
+                hits += Domain.get_names_by_prefix(name_prefix.replace('-', '.') + "-")
+
+            max_counter = 0
+            pattern = re.compile("^" + name_prefix + "-(\d+)$")
+            for hit in hits:
+                match = re.search(pattern, hit)
+                if match is not None:
+                    max_counter = max(max_counter, int(match.group(1)))
+            suffix = "-" + str(max_counter + 1)
+            name = name[:max_length - len(suffix)] + suffix
```

Comment:
i trust that this logic works but it would be great to pull it out into a utility function and write some tests